lib/src/main/java/com/squareup/cycler/RecyclerMutations.kt

```diff
@@ -111,6 +191,87 @@ fun <I : Any, S : I, V : View> Creator<I, S, V>.dragHandle(view: View) {
   }.setUpDragHandle(view)
 }
 
+private class SwipeState<T : Any, V : View>(
+  val viewHolder: Recycler.ViewHolder<V>,
+  var bindData: SwipeBindData<T>) {
+
+  /**
+   * @return true if the binding was successful. It is, the under-view should be painted.
+   */
+  fun bind(viewTranslation: Float): Boolean {
+    val valid = bindData.bind(viewTranslation)
+    if (valid) {
+      viewHolder.bind(0, bindData)
+    }
+    return valid
+  }
+}
+
+/**
+ * Passed as data for the [RowSpec.bind] defined in [MutationExtensionSpec.swipeUnderViewSpec].
+ */
+class SwipeBindData<T : Any>
+@VisibleForTesting
+internal constructor(
+  val dataItem: T,
+  val swipedView: View,
+  private val allowedDirections: EnumSet<SwipeDirection>
+) {
+
+  private val rightToLeft = swipedView.layoutDirection == View.LAYOUT_DIRECTION_RTL
+
+  /**
+   * It will tell if the [bind] was called successfully and there's a [direction] and [percentage]
+   * to provide to the developer's handler.
+   */
+  internal var isValid = false
+
+  /**
+   * The direction the swipe is happening at the moment.
+   * If both directions are allowed this might change for the same swipe.
+   */
+  lateinit var direction: SwipeDirection
+    internal set
+
+  /**
+   * Represents the percentage of swiping to the size. 0f <= percentage <= 1f.
+   * - 0f means the view is exactly in its normal place.
+   * - 1f means the view is completely swiped to the [direction].
+   */
+  var percentage: Float = 0f
+    internal set
+
+  /**
+   * @return true if the binding was successful: valid direction and percentage.
+   */
+  fun bind(viewTranslation: Float): Boolean {
+    val viewWidth = swipedView.width
+    if (viewWidth <= 0) {
+      return false
+    }
+    val swipingLeft = viewTranslation < 0
+    val candidateDirections = when {
+      swipingLeft && !rightToLeft -> leftDirections
+      !swipingLeft && !rightToLeft -> rightDirections
+      swipingLeft && rightToLeft -> leftDirectionsRtl
+      !swipingLeft && rightToLeft -> rightDirectionsRtl
+      else -> throw UnsupportedOperationException("Unreachable")
```

Comment:
Nit: This doesn't seem like an idiomatic use of UOE. This should probably either be IllegalStateException or IllegalArgumentException. Probably the latter since it seems like this case can only happen when the argument is invalid.